### ISSUE TITLE
Train YOLOv5 and YOLOv8 on multiple nodes having multiple GPUs using Training Jobs in AWS Sagemaker

### ISSUE THREAD

**Issue by Rishu18** [ultralytics/yolov5#11953]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have a large set of images and I wish to train the model efficiently. 

I am trying to train the model using the PyTorch estimator from Training Jobs in AWS Sagemaker. I am able to train the model on a single instance but I wish to train it with multiple instances so as to train it quickly. I have 4 available instances with 8 GPUs each. 

In the PyTorch estimator, I am using .sh script as an entrypoint. It has commands to clone the Ultralytics yolov5 repository, to install the requirements.txt and finally to call train.py. The final command looks like this:
python -m torch.distributed.run --nproc_per_nodes 8 train.py --cache --data data.yaml --weights yolov5x.pt

PyTorch estimator is called using:
pt_estimator = PyTorch(entry_point='trainer.sh', role=sagemaker_role, src=<path>, framework_version='2.0.0', py_version='py310', output_path=<path>, instance_type=<instance>, instance_count=1)
pt_estimator.fit(....)

Again, I have 4 available instances and I wish to use all of them using
pt_estimator = PyTorch(....., instance_count=4)

I searched for this but I could not find how to do it in Sagemaker Training Jobs and how to create entrypoint for the same. Please help me with this.

**Comment by glenn-jocher** [comment]
@Rishu18 to train YOLOv5 and YOLOv8 on multiple nodes with multiple GPUs using Training Jobs in AWS Sagemaker, you can utilize the `torch.distributed` module to enable distributed training. Here's an example of how you can modify your entrypoint script:

First, you need to modify your `trainer.sh` script to include the necessary distributed training commands. Replace the existing `python` command with the following:

```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --cache --data data.yaml --weights yolov5x.pt
```

By specifying `--nproc_per_node 8`, you are telling the `torch.distributed.run` module to utilize 8 GPUs per node.

To specify multiple instances, you will need to update the `instance_count` parameter in the `PyTorch` estimator. Set it to 4 to utilize all available instances:

```python
pt_estimator = PyTorch(entry_point='trainer.sh', role=sagemaker_role, src=<path>, framework_version='2.0.0', py_version='py310', output_path=<path>, instance_type=<instance>, instance_count=4)
```

With these modifications, your training job will utilize multiple instances, each with 8 GPUs, resulting in faster training.

Please note that using multiple nodes with multiple GPUs might require additional considerations such as data parallelism and synchronized batch normalization. You may find additional resources on the PyTorch official documentation helpful for more advanced distributed training techniques.

I hope this helps! If you have any further questions, please feel free to ask.

**Comment by Rishu18** [comment]
I tried this already but it is not helping. It seems like the model is getting trained on multiple instances separately and there is no communication among them and no benefit of multiple machines.

My entrypoint script trainer.sh has the command you mentioned, i.e.
`python -m torch.distributed.run --nproc_per_node 8 train.py --cache --data data.yaml --weights yolov5x.pt`

I changed instance_count to 4 as well,
`pt_estimator = PyTorch(entry_point='trainer.sh', role=sagemaker_role, src=<path>, framework_version='2.0.0', py_version='py310', output_path=<path>, instance_type=<instance>, instance_count=4)`

Even after this, training time is the same as what is used to take with one instance. There is no benefit of multiple instances.

Additionally, I changed the parameter 's3_data_distribution_type' to 'ShardedByS3Key' as well but this did not help as well and the training time is still the same as with single instance.

**Comment by glenn-jocher** [comment]
@Rishu18 the `torch.distributed.run` command in your `trainer.sh` script along with specifying `--nproc_per_node 8` and setting `instance_count` to 4 in the `PyTorch` estimator are the correct steps to enable distributed training on multiple instances with multiple GPUs.

However, it seems that the training process is not benefiting from the use of multiple instances. This could be due to a variety of reasons. Here are a few things you can check:

1. **Networking**: Ensure that the instances are properly connected and have sufficient bandwidth between them for efficient communication. Slow network connections can hinder distributed training performance.

2. **Data Loading**: Confirm that the data is evenly distributed across the multiple instances. Uneven data distribution can lead to imbalanced workloads and slower training.

3. **Data Preprocessing**: Make sure that any preprocessing steps, such as data augmentation or normalization, are consistent across all instances. Inconsistencies can affect model convergence and training time.

4. **Model Synchronization**: Check if your model is synchronized properly during training. This is especially important for models with Batch Normalization layers. You can either use synchronized Batch Normalization or wrap the model with `torch.nn.SyncBatchNorm` to ensure synchronization across instances.

5. **Learning Rate Scaling**: Adjust the learning rate for distributed training. You can try scaling the learning rate by the number of GPUs (`8 * 4 = 32`) to maintain the same effective learning rate as in the single instance training.

As always, ensure that you have the latest version of both PyTorch and YOLOv5, and that all the required dependencies are installed correctly.

Hopefully, these tips will help improve the performance of your distributed training setup. If you have any further questions or concerns, please don't hesitate to ask.
